### PR TITLE
buildd: check internet after network command fails (CRAFT-1461)

### PR DESCRIPTION
**Pull request by facundobatista**
For this I centralized the command run in a helper, that simplifies all the calls and makes the verification easier. Note that I have another PR in mind to simplify the code more.

I tested this IRL, the resulting exception is much more clear than the obscure failings we had in the past:
```
2022-11-11 13:57:44.915 charmcraft internal error: NetworkError(brief='A network related operation failed in a context of no network access.', details=None, resolution='Verify that the environment has internet connectivity.')                                                                                                                                                                                  
2022-11-11 13:57:44.917 Traceback (most recent call last):                                                                                                                                               
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/craft_providers/bases/buildd.py", line 114, in _execute_run                                                                                      
2022-11-11 13:57:44.917     proc = executor.execute_run(command, check=check, capture_output=capture_output, text=text)                                                                                  
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/craft_providers/lxd/lxd_instance.py", line 289, in execute_run                                                                                   
2022-11-11 13:57:44.917     return self.lxc.exec(                                                                                                                                                        
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/craft_providers/lxd/lxc.py", line 329, in exec                                                                                                   
2022-11-11 13:57:44.917     return runner(final_cmd, **kwargs)  # pylint: disable=subprocess-run-check                                                                                                   
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/usr/lib/python3.8/subprocess.py", line 516, in run                                                                                                   
2022-11-11 13:57:44.917     raise CalledProcessError(retcode, process.args,                                                                                                                              
2022-11-11 13:57:44.917 subprocess.CalledProcessError: Command '['lxc', '--project', 'charmcraft', 'exec', 'local:charmcraft-my-super-charm-27791466-0-0-amd64', '--', 'env', 'PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin', 'CHARMCRAFT_MANAGED_MODE=1', 'apt-get', 'update']' returned non-zero exit status 100.                                                                 
2022-11-11 13:57:44.917                                                                                                                                                                                  
2022-11-11 13:57:44.917 The above exception was the direct cause of the following exception:                                                                                                             
2022-11-11 13:57:44.917 Traceback (most recent call last):                                                                                                                                               
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/charmcraft/main.py", line 184, in main                                                                                                           
2022-11-11 13:57:44.917     retcode = dispatcher.run()                                                                                                                                                   
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/craft_cli/dispatcher.py", line 448, in run                                                                                                       
2022-11-11 13:57:44.917     return self._loaded_command.run(self._parsed_command_args)                                                                                                                   
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/charmcraft/commands/pack.py", line 133, in run                                                                                                   
2022-11-11 13:57:44.917     pack_method(parsed_args)                                                                                                                                                     
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/charmcraft/commands/pack.py", line 165, in _pack_charm                                                                                           
2022-11-11 13:57:44.917     charms = builder.run(                                                                                                                                                        
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/charmcraft/instrum.py", line 152, in _f                                                                                                          
2022-11-11 13:57:44.917     return func(*args, **kwargs)                                                                                                                                                 
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/charmcraft/commands/build.py", line 292, in run                                                                                                  
2022-11-11 13:57:44.917     charm_name = self.pack_charm_in_instance(                                                                                                                                    
2022-11-11 13:57:44.917   File "/snap/charmcraft/x2/lib/charmcraft/commands/build.py", line 357, in pack_charm_in_instance                                                                               
2022-11-11 13:57:44.918     with self.provider.launched_environment(                                                                                                                                     
2022-11-11 13:57:44.918   File "/snap/charmcraft/x2/usr/lib/python3.8/contextlib.py", line 113, in __enter__                                                                                             
2022-11-11 13:57:44.918     return next(self.gen)                                                                                                                                                        
2022-11-11 13:57:44.918   File "/snap/charmcraft/x2/lib/craft_providers/lxd/lxd_provider.py", line 126, in launched_environment                                                                          
2022-11-11 13:57:44.918     instance = launch(                                                                                                                                                           
2022-11-11 13:57:44.918   File "/snap/charmcraft/x2/lib/craft_providers/lxd/launcher.py", line 212, in launch                                                                                            
2022-11-11 13:57:44.918     base_configuration.setup(executor=instance)                                                                                                                                  
2022-11-11 13:57:44.918   File "/snap/charmcraft/x2/lib/craft_providers/bases/buildd.py", line 392, in setup                                                                                             
2022-11-11 13:57:44.918     self._setup_apt(executor=executor, deadline=deadline)                                                                                                                        
2022-11-11 13:57:44.918   File "/snap/charmcraft/x2/lib/craft_providers/bases/buildd.py", line 556, in _setup_apt                                                                                        
2022-11-11 13:57:44.918     _execute_run(executor, ["apt-get", "update"], verify_network=True)                                                                                                           
2022-11-11 13:57:44.918   File "/snap/charmcraft/x2/lib/craft_providers/bases/buildd.py", line 117, in _execute_run                                                                                      
2022-11-11 13:57:44.918     raise NetworkError() from exc                                                                                                                                                
2022-11-11 13:57:44.918 craft_providers.bases.errors.NetworkError: A network related operation failed in a context of no network access.                                                                 
2022-11-11 13:57:44.918 Verify that the environment has internet connectivity.                                                                                                                           
2022-11-11 13:57:44.918 Full execution log: '/home/facundo/snap/charmcraft/common/cache/charmcraft/log/charmcraft-20221111-135607.196724.log'                      
```

That said, we will talk about Craft Provider errors not being "crashes" but "failures", but that discussion exceeds this PR.

### CODE REVIEW

File: craft_providers/bases/buildd.py [hunk 1]
```diff
@@ -78,6 +78,57 @@ def _check_deadline(
         raise BaseConfigurationError(brief=message)
 
 
+def _network_connected(executor):
+    """Check if the network is connected."""
+    # check if the port is open using bash's built-in tcp-client (1.1.1.1 is a well
+    # known DNS server, 53 is the DNS port)
+    command = ["bash", "-c", "exec 3<> /dev/tcp/1.1.1.1/53"]
```
Comment (sergiusens) [inline craft_providers/bases/buildd.py:85]: 1.1.1.1 will be blocked in many enterprises which is why we went with getent https://github.com/canonical/craft-providers/pull/176/files#diff-bcc16e27108e1d381f59fe92f1439c56ac920252ab7be057491af4a20bcb7748R817

**Review by sergiusens — COMMENTED** [review]
checking for 1.1.1.1 will not work in strict enterprises

please make the commit message shorter

**Review by sergiusens — CHANGES_REQUESTED** [review]
checking for 1.1.1.1 will not work in strict enterprises

please make the commit message shorter

File: craft_providers/bases/buildd.py (hunk 1 above)
Comment (facundobatista) [reply craft_providers/bases/buildd.py:85]: You're right about `1.1.1.1` maybe blocked. 

I don't like the `getent` solution because it's too indirect, some systems may not have that mechanism, and mainly because we're currently using it to assert that the network is already there before continuing with the setup, which then fails because of missing network.

After a conversation with Igor and Callahan we decided that we could improve this verification by actually referring our site, using the HTTPS port.

File: craft_providers/bases/buildd.py [hunk 2]
```diff
@@ -827,12 +814,9 @@ def _setup_wait_for_network(
         logger.debug("Waiting for networking to be ready...")
 
         _check_deadline(deadline)
+        command = ["getent", "hosts", "snapcraft.io"]
```
Comment (mr-cal) [inline craft_providers/bases/buildd.py:817]: pedantic: It would be nice if this network check used the same bash tcp call as the new one

**Review by mr-cal — COMMENTED** [review]
+1.

I built snapcraft with this branch and disconnected the network during the instance setup.  Works very well.  It will be even more useful once we start using `_check_deadline()`, since some of the operations take a long time to time out (`apt-get update` takes 3.5 minutes to timeout).

File: craft_providers/bases/buildd.py [hunk 3]
```diff
@@ -78,6 +78,57 @@ def _check_deadline(
         raise BaseConfigurationError(brief=message)
 
 
+def _network_connected(executor):
+    """Check if the network is connected."""
+    # check if the port is open using bash's built-in tcp-client, communicating with
+    # the to HTTPS port on our site
+    command = ["bash", "-c", "exec 3<> /dev/tcp/snapcraft.io/443"]
```
[the remark this answers is not in the document]
Comment (facundobatista) [reply craft_providers/bases/buildd.py:85]: It looks like it's not. I couldn't make it go through a proxy, and [this post](https://unix.stackexchange.com/questions/288379/using-dev-tcp-with-proxy) says that it does not (for an old bash, but as we also support old systems...).

We could do this verification only if `http_proxy` and `https_proxy` are not set...

[the remark this answers is not in the document]
Comment (simondeziel) [reply craft_providers/bases/buildd.py:92]: It is common for `http_proxy` (note no `s`) to allow connecting to HTTPS destinations using the [CONNECT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT) method.

**Review by mr-cal — APPROVED** [review]
+1.  Changes since my last review look good and I like the extra comments.

File: craft_providers/bases/buildd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (simondeziel) [reply craft_providers/bases/buildd.py:92]: Sorry, I got confused by the fact that `https_proxy` can use a proxy URL with HTTP to then use the CONNECT method.
tl;dr: your code is fine, sorry for the noise!

**Comment by facundobatista** [comment]
Superseded by #189
